add ability to email payloads in bulk

mkangia opened the pull request:
[Spec](https://docs.google.com/document/d/1N1cLHi3CDVAiTkm1xnDEfO6wBgbqdTHDccHjALDXhwg/edit#heading=h.29nthshn87qv)

Needs some refinement but core implementation in place.

buddy: @snopoke

mkangia committed:
add ability to email payloads in bulk

esoergel commented on corehq/apps/domain/utils.py:
```diff
@@ -111,3 +115,62 @@ def guess_domain_language(domain_name):
+            payload_id=payload_id,
+        )
+        payload_doc = repeater.payload_doc(dummy_repeat_record)
+        payload = Repeater(None).get_payload(dummy_repeat_record, payload_doc)
```
shouldn't this be `repeater` instead of `Repeater(None)`?

esoergel commented on corehq/apps/domain/utils.py:
```diff
@@ -111,3 +115,62 @@ def guess_domain_language(domain_name):
+                payloads[payload_id] = {'note': 'Could not generate payload, %s' % e.message}
+
+    def create_result_file():
+        with open(result_file_name, 'w') as csvfile:
```
Is this typically how we handle tempfiles?  I thought they were stored somewhere in particular - this will save it in the root of the repo.

esoergel replied:
It looks like `DownloadBase` uses a tempfile:
https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/soil/__init__.py#L67
```python
In [1]: from tempfile import mkstemp

In [2]: mkstemp(suffix='.txt')
Out[2]: (10, '/tmp/tmpEAk0C8.txt')
```
I'm not totally sure of the distinction, except that tempfiles are sometimes wiped, and that files in the repo root are wiped on deploy.  I'd go with whatever `DownloadBase` uses, since that's been vetted.  Could be better if we run into disk space issues, and it may be safer if this occurs during a deploy.

esoergel commented on corehq/apps/domain/views.py:
```diff
@@ -564,6 +565,21 @@ def requeue_repeat_record(request, domain):
+        repeater_id = request.POST.get('repeater_id')
+        data = csv.reader(request.FILES['payload_ids_file'])
+        payload_ids = [row[0] for row in data]
+        send_mock_repeater_payloads(repeater_id, payload_ids, email_id)
```
Might wanna make this an async task, particularly since it doesn't return anything in this view.

esoergel commented on corehq/apps/domain/views.py:
```diff
@@ -564,6 +565,21 @@ def requeue_repeat_record(request, domain):
+        payload_ids = [row[0] for row in data]
+        send_mock_repeater_payloads(repeater_id, payload_ids, email_id)
+        messages.success(request, _("Successfully emailed payloads."))
+    except Exception as e:
```
I think this is overbroad exception catching.  In particular, `send_mock_repeater_payloads` seems like it shouldn't be part of this `try/except` block.

mkangia replied:
Its just that we are accepting a file from the user and it is possible that its not in expected/parsable format or there is some unexpected error. So letting the user know we could not parse the file seems fine here.
Agree on `send_mock_repeater_payloads`.

esoergel reviewed (commented):
Had a couple comments, but looks good overall.  One bit of qualitative feedback: I found the word "mock" here to be misleading - these are real payloads.

mkangia committed:
review: rename, make report a background task and rescue only where needed

mkangia commented:
@esoergel thanks for your last review. Have made progress. Can you please review this once again?

esoergel reviewed (approved):
Nice, thanks for the follow-up!

mkangia committed:
lint

mkangia committed:
update file name format to avoid issues with saving files with some systems

have seen these earlier with a name like bulk-payloads-2018-02-20 16:42:54.488399 having issues with : in file name

mkangia committed:
remove file post email [skip ci]

mkangia committed:
Merge master: Resolve conflicts

mkangia commented on corehq/apps/domain/utils.py:
```diff
@@ -111,3 +116,67 @@ def guess_domain_language(domain_name):
+            email.attach(filename=result_file_name, content=csvfile.read())
+            csvfile.close()
+            email.send()
+            os.remove(result_file_name)
```
@esoergel 
thought about this a bit. You suggested using a Temp file..
If i just all content in memory and then just email it, the only concern was if the content is quite large, it would take quite some space in memory. So kept it in the file only and removed it post email.
Let me know if there are any concerns with this.

esoergel replied:
I was actually just looking at this same sort of thing with Jordan.  The locations export is the simplest example I saw: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/locations/util.py#L283-L302
This uses a temporary file location to stream to while writing the file, then saves it in Riak for a timeout period and provides a location where the user can download it.  The nice thing about that approach is that you don't have to send any potentially sensitive info over email.

It depends on what this is going to be used for though.  This is probably fine as long as you know the contents of the file aren't going to contain sensitive info.  If they might, we should make an effort to keep that to secured channels.

esoergel reviewed (commented):
I'm hesitant about emailing payload bodies, but otherwise this looks good.

mkangia committed:
create a temp downloadable file instead of sending it in the email.

mkangia committed:
remove extra line added

mkangia committed:
add back line

mkangia committed:
Merge branch 'master' into mk-mock-repeater-report

mkangia committed:
add code removed during merge conflict resolution

Modal and custom template to be used

mkangia committed:
fix issue with arguments and file picked for zip

mkangia commented:
@esoergel updated this to email a link to a file that can be downloaded now. Let me know your thoughts.. this is also deployed on staging if you want to take a look or try it out..

esoergel commented:
Sounds great, should be good to go now, yeah?

mkangia committed:
Merge branch 'master' into mk-mock-repeater-report

mkangia committed:
check on attribute only when present

mkangia committed:
extract ExposeBlobDownload

mkangia committed:
lint

mkangia commented:
@esoergel thank you for the suggestion. Extracted out the utility.

Unsure about the build failure. Do you think its related to the changes?

esoergel commented on corehq/apps/domain/utils.py:
```diff
@@ -176,46 +172,13 @@ def create_result_file():
-    clean_temp_files(temp_file_path, temp_zip_path)
-    email_result(download_id)
+    download_url = ExposeBlobDownload().get_link(temp_file_path, result_file_name, Format.CSV)
+    email_result(download_url)
```
awesome

esoergel reviewed (approved):
I'm guessing the build failure is a fluke, it looked like some JS thing related to ICDS.  I restarted the build.

mkangia committed:
fix method call

mkangia committed:
fix error with headers variable being accessed before assignment

mkangia committed:
remove unused import

mkangia commented:
thanks @esoergel 
Tested this once quickly on staging and made a minor fix. Gtg!
